test/unit/FixedLatencyMemoryInterfaceTest.cc

```diff
@@ -0,0 +1,122 @@
+#include "gtest/gtest.h"
+#include "simeng/FixedLatencyMemoryInterface.hh"
+
+namespace {
+
+class FixedLatencyMemoryInterfaceTest : public testing::Test {
+ public:
+  FixedLatencyMemoryInterfaceTest()
+      : memory(memoryData.data(), memorySize, latency) {}
+
+ protected:
+  static constexpr uint16_t memorySize = 4;
+  const uint16_t latency = 2;
+  std::array<char, memorySize> memoryData = {(char)0xFE, (char)0xCA, (char)0xBA,
+                                             (char)0xAB};
+
+  simeng::RegisterValue value = {0xDEADBEEF, 4};
+  simeng::RegisterValue value_oversized = {0xDEADBEEFDEADBEEF, 8};
+  simeng::MemoryAccessTarget target = {0, 4};
+  simeng::MemoryAccessTarget target_OutOfBound1 = {1000, 4};
+  simeng::MemoryAccessTarget target_OutOfBound2 = {0, 8};
+
+  const std::string writeOverflowStr =
+      "Attempted to write beyond memory limit.";
+
+  simeng::FixedLatencyMemoryInterface memory;
+};
+
+// Test that we can read data and it completes after a number of cycles.
+TEST_F(FixedLatencyMemoryInterfaceTest, FixedReadData) {
+  // Read a 32-bit value
+  memory.requestRead(target, 1);
+  EXPECT_TRUE(memory.hasPendingRequests());
+
+  // Tick once - request should still be pending
+  memory.tick();
+  EXPECT_TRUE(memory.hasPendingRequests());
+
+  // Tick again - request should have completed
+  memory.tick();
+  EXPECT_FALSE(memory.hasPendingRequests());
+
+  auto entries = memory.getCompletedReads();
+  EXPECT_EQ(entries.size(), 1);
+  EXPECT_EQ(entries[0].requestId, 1);
+  EXPECT_EQ(entries[0].data, simeng::RegisterValue(0xABBACAFE, 4));
+  EXPECT_EQ(entries[0].target, target);
+}
+
+// Test that we can write data and it completes after a number of cycles.
+TEST_F(FixedLatencyMemoryInterfaceTest, FixedWriteData) {
+  // Write a 32-bit value to memory
+  memory.requestWrite(target, value);
+  EXPECT_TRUE(memory.hasPendingRequests());
+
+  // Tick once - request should still be pending
+  memory.tick();
+  EXPECT_TRUE(memory.hasPendingRequests());
+
+  // Tick again - request should have completed
+  memory.tick();
+  EXPECT_FALSE(memory.hasPendingRequests());
+  EXPECT_EQ(reinterpret_cast<uint32_t*>(memoryData.data())[0], 0xDEADBEEF);
+}
+
+// Test that out-of-bounds memory reads are correctly handled.
+TEST_F(FixedLatencyMemoryInterfaceTest, OutofBoundsRead) {
+  // Create a target such that address + size will overflow
+  memory.requestRead(target_OutOfBound1, 1);
+
+  // Create a regular out-of-bounds target
+  memory.requestRead(target_OutOfBound2, 2);
+
+  // Tick once - request shouldn't have completed
+  memory.tick();
+  EXPECT_TRUE(memory.hasPendingRequests());
+
+  // Tick twice - request should have completed
+  memory.tick();
+  EXPECT_FALSE(memory.hasPendingRequests());
+
+  auto entries = memory.getCompletedReads();
+  EXPECT_EQ(entries.size(), 2);
+
+  auto overflowResult = entries[0];
+  EXPECT_EQ(overflowResult.requestId, 1);
+  EXPECT_FALSE(overflowResult.data);
```

Comment:
Given we have the equality operator defined now, can't we keep this as `EXPECT_EQ(overflowResult.data, simeng::RegisterValue());`? Not sure what the `EXPECT_FALSE` is doing regarding trying to equate a class to false

Reply:
Due to the assertion `assert(bytes > 0 && "Attempted to access an uninitialised RegisterValue");` in the RegisterValue class, we cannot use the equality operator. And there is not a way I can see to check the equality of two register values without getting the underlying data by using `getAsVector`.

By using `EXPECT_FALSE`, we are mimicing the logic found in `Instruction::supplyData` to check for a data abort (i.e. `if (!data)`) which checks for uninitialised Register Values

Reply:
Okay, looks good to me then